src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/KeyedDependencyInjectionSpecificationTests.cs

```diff
@@ -324,6 +324,41 @@ public void ResolveKeyedServiceTransientTypeWithAnyKey()
             Assert.NotSame(first, second);
         }
 
+        [Fact]
+        public void ResolveKeyedServiceFromInjectedServiceProvider()
+        {
+            var service1 = new Service();
+            var service2 = new Service();
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService>("service1", service1);
+            serviceCollection.AddKeyedSingleton<IService>("service2", service2);
+            serviceCollection.AddSingleton<ServiceProviderAccessor>();
+
+            var provider = CreateServiceProvider(serviceCollection);
+            var accessor = provider.GetRequiredService<ServiceProviderAccessor>();
+
+            Assert.Null(accessor.ServiceProvider.GetService<IService>());
+            Assert.Same(service1, accessor.ServiceProvider.GetKeyedService<IService>("service1"));
+            Assert.Same(service2, accessor.ServiceProvider.GetKeyedService<IService>("service2"));
+        }
+
+        [Fact]
+        public void ResolveKeyedServiceFromScopeServiceProvider()
+        {
+            var service1 = new Service();
+            var service2 = new Service();
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedSingleton<IService>("service1", service1);
+            serviceCollection.AddKeyedSingleton<IService>("service2", service2);
+
+            var provider = CreateServiceProvider(serviceCollection);
+            var scope = provider.GetRequiredService<IServiceScopeFactory>().CreateScope();
```

Comment:
Nit: Can we test with scoped services and different scopes to make sure we're not incorrectly getting the same instances from the IServiceScopeFactory?